numpy/core/src/multiarray/dtypemeta.c

```diff
@@ -400,9 +400,10 @@ default_builtin_common_dtype(PyArray_DTypeMeta *cls, PyArray_DTypeMeta *other)
 static PyArray_DTypeMeta *
 string_unicode_common_dtype(PyArray_DTypeMeta *cls, PyArray_DTypeMeta *other)
 {
-    assert(cls->type_num < NPY_NTYPES);
+    assert(cls->type_num < NPY_NTYPES && cls != other);
```

Comment:
I am currently guaranteeing this, if that would change UNICODE+UNICODE would misbehave (but that would also lead to very loud errors all over the place).